File: modules/aws_scanner/ec2/ec2_scanner.py
import boto3
import json
import uuid
import time
import logging
from datetime import datetime
from botocore.exceptions import ClientError

# Set up logger
logger = logging.getLogger("EC2Scanner")
logger.setLevel(logging.INFO)

# Ensure log directories exist (relative to current directory)
import os

# Define log paths relative to project root
PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))  # modules/aws_scanner/ec2/../..
LOG_DIR = os.path.join(PROJECT_ROOT, "logs", "merope")
SCAN_LOG_DIR = os.path.join(PROJECT_ROOT, "logs", "aws_scans", "ec2")

os.makedirs(LOG_DIR, exist_ok=True)              # Create logs/merope if not exists
os.makedirs(SCAN_LOG_DIR, exist_ok=True)          # Create logs/aws_scans/ec2 if not exists

# Add file handler for logging
log_file_path = os.path.join(LOG_DIR, "scanner.log")
handler = logging.FileHandler(log_file_path)
handler.setFormatter(logging.Formatter('%(asctime)s [%(levelname)s] %(message)s'))
logger.addHandler(handler)

# Add console handler
console_handler = logging.StreamHandler()
console_handler.setFormatter(handler.formatter)
logger.addHandler(console_handler)

# Custom JSON Encoder to handle datetime objects
class CustomJSONEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, datetime):
            return obj.isoformat()  # Convert datetime to ISO string
        elif isinstance(obj, bytes):
            return obj.decode('utf-8', errors='ignore')  # Handle binary data
        return super().default(obj)

# Generate unique scan ID
def generate_scan_id():
    return f"SCAN-{uuid.uuid4()}-{int(time.time())}"

# Discover active AWS regions
def get_active_regions(session):
    try:
        ec2_client = session.client('ec2', region_name='us-east-1')
        response = ec2_client.describe_regions()
        return [region['RegionName'] for region in response['Regions']]
    except Exception as e:
        logger.warning(f"⚠️ Could not fetch active regions: {e}")
        return ['us-east-1']

# Check if instance is orphaned
def is_orphaned(instance):
    return len(instance.get('Tags', [])) == 0

# Get associated resources (Security Groups, Volumes, ENIs)
def get_associated_resources(ec2, instance):
    result = {
        "security_groups": [],
        "volumes": [],
        "network_interfaces": []
    }

    sg_ids = [sg['GroupId'] for sg in instance.get('SecurityGroups', [])]
    if sg_ids:
        try:
            result['security_groups'] = ec2.describe_security_groups(GroupIds=sg_ids).get('SecurityGroups', [])
        except Exception as e:
            logger.error(f"🚨 Error fetching security groups: {e}")

    try:
        result['volumes'] = ec2.describe_volumes(Filters=[{
            'Name': 'attachment.instance-id',
            'Values': [instance['InstanceId']]
        }]).get('Volumes', [])
    except Exception as e:
        logger.error(f"🚨 Error fetching volumes: {e}")

    try:
        result['network_interfaces'] = ec2.describe_network_interfaces(Filters=[{
            'Name': 'attachment.instance-id',
            'Values': [instance['InstanceId']]
        }]).get('NetworkInterfaces', [])
    except Exception as e:
        logger.error(f"🚨 Error fetching ENIs: {e}")

    return result

# Send message to RabbitMQ
def send_to_rabbitmq(message):
    try:
        import pika

        connection = pika.BlockingConnection(pika.ConnectionParameters(
            host='localhost'
        ))
        channel = connection.channel()
        channel.queue_declare(queue='scan_queue', durable=True)
        channel.basic_publish(
            exchange='',
            routing_key='scan_queue',
            body=json.dumps(message, cls=CustomJSONEncoder),
            properties=pika.BasicProperties(delivery_mode=2),  # Persistent message
            mandatory=True
        )
        connection.close()
        logger.info(f"[x] Sent EC2 Instance: {message['resource_id']}")
    except Exception as e:
        logger.error(f"❌ Failed to send to RabbitMQ: {e}")

# Scan EC2 instances in one region
def scan_ec2_instances(session, scan_id, region):
    logger.info(f"🌍 Scanning EC2 in {region}")
    results = []

    try:
        ec2 = session.client('ec2', region_name=region)
        paginator = ec2.get_paginator('describe_instances')
        page_iterator = paginator.paginate()

        for page in page_iterator:
            reservations = page.get('Reservations', [])
            logger.info(f"📦 Found {len(reservations)} instances in {region}")

            for reservation in reservations:
                for instance in reservation.get('Instances', []):
                    if instance['State']['Name'] != 'running':
                        continue

                    item = {
                        "scan_id": scan_id,
                        "region": region,
                        "service": "ec2",
                        "resource_type": "instance",
                        "resource_id": instance['InstanceId'],
                        "data": instance,
                        "orphaned": is_orphaned(instance),
                        "public_ip": 'PublicIpAddress' in instance,
                        "timestamp": time.time(),
                        "missing_metadata": {
                            "tags_missing": len(instance.get('Tags', [])) == 0
                        },
                        "associated_resources": get_associated_resources(ec2, instance)
                    }

                    # Save raw scan to file
                    scan_log_path = os.path.join(SCAN_LOG_DIR, f"{instance['InstanceId']}.json")
                    with open(scan_log_path, 'w') as f:
                        json.dump(item, f, indent=2, cls=CustomJSONEncoder)

                    # Send to RabbitMQ
                    send_to_rabbitmq(item)
                    results.append(item)

        return results

    except ClientError as ce:
        logger.warning(f"🚫 EC2 not supported in {region}: {ce}")
        return []

# Main entry point
if __name__ == "__main__":
    logger.info("🔌 Initializing AWS Session...")
    session = boto3.Session()
    scan_id = generate_scan_id()
    regions = get_active_regions(session)

    all_results = []

    for region in regions:
        logger.info(f"🔍 Scanning region: {region}")
        results = scan_ec2_instances(session, scan_id, region)
        all_results.extend(results)

    logger.info(f"✅ Sent {len(all_results)} running EC2 instances.")

File: modules/scanner/ec2_scanner.go
package main

import (
    "context"
    "encoding/json"
    "fmt"
    "log"
    "os"
    "time"

    "github.com/aws/aws-sdk-go-v2/config"
    "github.com/aws/aws-sdk-go-v2/service/ec2"
    "github.com/confluentinc/confluent-kafka-go/v2/kafka"
)

type Instance struct {
    ID               string                 `json:"resource_id"`
    Region           string               `json:"region"`
    Service          string               `json:"service"`
    ResourceType     string               `json:"resource_type"`
    ScanID           string               `json:"scan_id"`
    Data             map[string]interface{} `json:"data"`
    Orphaned         bool                 `json:"orphaned"`
    PublicIP         bool                 `json:"public_ip"`
    MissingMetadata  map[string]bool      `json:"missing_metadata"`
    AssociatedResources map[string]interface{} `json:"associated_resources"`
    Timestamp        int64                `json:"timestamp"`
}

var logger *log.Logger

func init() {
    logFile, _ := os.OpenFile("/logs/merope/scanner.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
    logger = log.New(logFile, "[EC2Scanner] ", log.LstdFlags)
}

func convertInstance(instance *ec2.Instance) map[string]interface{} {
    m := make(map[string]interface{})
    m["InstanceId"] = instance.InstanceId
    m["State"] = instance.State.Name
    m["PublicIpAddress"] = instance.PublicIpAddress
    m["Tags"] = instance.Tags
    return m
}

func isOrphaned(instance map[string]interface{}) bool {
    tags, ok := instance["Tags"].([]interface{})
    return !ok || len(tags) == 0
}

func scanEC2Instances(scanID string) ([]Instance, error) {
    cfg, _ := config.LoadDefaultConfig(context.TODO())
    svc := ec2.NewFromConfig(cfg)

    input := &ec2.DescribeInstancesInput{}
    paginator := ec2.NewDescribeInstancesPaginator(svc, input)

    var results []Instance

    for paginator.HasMorePages() {
        page, err := paginator.NextPage(context.TODO())
        if err != nil {
            log.Printf("🚨 Error fetching page: %v", err)
            continue
        }

        for _, reservation := range page.Reservations {
            for _, instance := range reservation.Instances {
                instanceMap := convertInstance(instance)

                results = append(results, Instance{
                    ID:              *instance.InstanceId,
                    Region:          *svc.Options().Region,
                    Service:         "ec2",
                    ResourceType:    "instance",
                    ScanID:          scanID,
                    Data:            instanceMap,
                    Orphaned:        isOrphaned(instanceMap),
                    PublicIP:        instance.PublicIpAddress != nil,
                    MissingMetadata: map[string]bool{"tags_missing": isOrphaned(instanceMap)},
                    AssociatedResources: map[string]interface{}{
                        "security_groups": instance.SecurityGroups,
                        "vpc_id":        instance.VpcId,
                    },
                    Timestamp: time.Now().Unix(),
                })

                sendToKafka(results[len(results)-1])
                saveToDisk(*instance.InstanceId, results[len(results)-1])
            }
        }
    }

    return results, nil
}

func saveToDisk(id string, item Instance) {
    os.MkdirAll("/logs/aws_scans/ec2", os.ModePerm)
    jsonData, _ := json.MarshalIndent(item, "", "  ")
    os.WriteFile(fmt.Sprintf("/logs/aws_scans/ec2/%s.json", id), jsonData, 0644)
}

func sendToKafka(item Instance) {
    p, err := kafka.NewProducer(&kafka.ConfigMap{"bootstrap.servers": "kafka:9092"})
    if err != nil {
        log.Fatalf("❌ Failed to create Kafka producer: %v", err)
    }
    defer p.Close()

    jsonData, _ := json.Marshal(item)
    deliveryChan := make(chan kafka.Event)

    err = p.Produce(&kafka.Message{
        TopicPartition: kafka.TopicPartition{Topic: "aws_scan_data", Partition: kafka.PartitionAny},
        Key:            []byte(item.ID),
        Value:          jsonData,
    }, deliveryChan)

    e := <-deliveryChan
    msg := e.(*kafka.Message)
    if msg.TopicPartition.Error != nil {
        log.Printf("❌ Kafka delivery failed: %v", msg.TopicPartition.Error)
    } else {
        log.Printf("✅ Sent %s | Offset: %v", item.ID, msg.TopicPartition.Offset)
    }
    close(deliveryChan)
}

func getActiveRegions(session *ec2.Client) ([]string, error) {
    output, err := session.DescribeRegions(context.TODO(), &ec2.DescribeRegionsInput{})
    if err != nil {
        return []string{"us-east-1"}, nil
    }

    var regions []string
    for _, r := range output.Regions {
        regions = append(regions, *r.RegionName)
    }
    return regions, nil
}

func main() {

    fmt.Println("📡 Starting Merope AWS Scanner v1.0")
    scanID := fmt.Sprintf("SCAN-%d", time.Now().UnixNano())

    log.Println("🔌 Initializing AWS Session...")
    session := ec2.NewFromConfig(config.LoadDefaultConfig(context.TODO()))

    regions, _ := getActiveRegions(session)
    for _, region := range regions {
        client := ec2.NewFromConfig(config.LoadDefaultConfig(context.TODO(), config.WithRegion(region)))
        fmt.Printf("🌍 Scanning region: %s\n", region)

        scanner := func() ([]Instance, error) {
            return scanEC2Instances(scanID)
        }

        instances, _ := scanner()
        log.Printf("📦 Found %d running EC2 instances in %s\n", len(instances), region)
    }
}
